cli/cmd/awsiam.go

```diff
@@ -0,0 +1,324 @@
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	// Only destroy the instance profile if it's ours
+	if *instanceProfile.InstanceProfileName == instanceProfileName {
+		// Remove role from instance profile
+		cli.Log.Debugw("removing role from instance profile", "role", role, "instance profile", instanceProfile)
+		removeInput := &iam.RemoveRoleFromInstanceProfileInput{
+			InstanceProfileName: instanceProfile.InstanceProfileName,
+			RoleName:            role.RoleName,
+		}
+		_, err := c.RemoveRoleFromInstanceProfile(context.Background(), removeInput)
+		if err != nil {
+			return err
+		}
+
+		// Delete instance profile
+		cli.Log.Debugw("deleting instance profile", "instance profile", instanceProfile)
+		deleteProfileInput := &iam.DeleteInstanceProfileInput{
+			InstanceProfileName: instanceProfile.InstanceProfileName,
+		}
+		_, err = c.DeleteInstanceProfile(context.Background(), deleteProfileInput)
+		if err != nil {
+			return err
+		}
+	}
+
+	if byoRoleName != "" || *role.RoleName != roleName {
+		return nil // we didn't create this role, we should not delete it
+	}
+
+	// List managed policies attached to this role (assume there are no inline policies)
+	listInput := &iam.ListAttachedRolePoliciesInput{
+		RoleName: role.RoleName,
+	}
+	listOutput, err := c.ListAttachedRolePolicies(context.Background(), listInput)
+	if err != nil {
+		return err
+	}
+
+	// Detach managed policies
+	for _, attachedPolicy := range listOutput.AttachedPolicies {
+		cli.Log.Debugw("detaching policy", "policy", attachedPolicy, "role", role)
+		detachInput := &iam.DetachRolePolicyInput{
+			PolicyArn: attachedPolicy.PolicyArn,
+			RoleName:  role.RoleName,
+		}
+		_, err := c.DetachRolePolicy(context.Background(), detachInput)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete the role
+	cli.Log.Debugw("deleting role", "role", role)
+	deleteRoleInput := &iam.DeleteRoleInput{
+		RoleName: role.RoleName,
+	}
+	_, err = c.DeleteRole(context.Background(), deleteRoleInput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setupSSMRole sets up an IAM role for SSM to assume.
+// If `roleName` is not the empty string, then use that role
+// instead of creating a new one.
+func setupSSMRole(cfg aws.Config, roleName string) (types.Role, error) {
+	if roleName != "" {
+		return getRoleFromName(cfg, roleName)
+	} else { // user did not provide a role, creating one now
+		return createSSMRole(cfg)
+	}
+}
+
+func getRoleFromName(cfg aws.Config, roleName string) (types.Role, error) {
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	input := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
+	}
+	output, err := c.GetRole(context.Background(), input)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	return *output.Role, nil
+}
+
+// createSSMRole makes a call to the AWS API to create an IAM role.
+// This role allows the current user to assume it.
```

Comment:
I might be missing something but I don't see how this role is scoped such that *only* the current user can assume this role. We need to make sure that only the current user can assume the role...

Reply:
Good catch, this comment was when I (mistakenly) thought that the role was meant for the current user to assume. Actually, it's for the EC2 service to assume. I updated this with 16643d2428e68ecffa310d251596b1bb5990066b